drivers/xbee/xbee.c

```diff
@@ -0,0 +1,700 @@
+    dev->sleep_pin = sleep_pin;
+    /* set default options */
+    dev->proto = XBEE_DEFAULT_PROTOCOL;
+    dev->options = 0;
+    dev->addr_short[0] = (uint8_t)(XBEE_DEFAULT_SHORT_ADDR >> 8);
+    dev->addr_short[1] = (uint8_t)(XBEE_DEFAULT_SHORT_ADDR);
+    /* initialize buffers and locks*/
+    mutex_init(&(dev->tx_lock));
+    mutex_init(&(dev->resp_lock));
+    dev->resp_limit = 1;    /* needs to be greater then 0 initially */
+    dev->rx_count = 0;
+    /* initialize UART and GPIO pins */
+    if (uart_init(uart, baudrate, _rx_cb, _tx_cb, dev) < 0) {
+        DEBUG("xbee: Error initializing UART\n");
+        return -ENXIO;
+    }
+    if (reset_pin < GPIO_NUMOF) {
+        if (gpio_init_out(reset_pin, GPIO_NOPULL) < 0) {
+            DEBUG("xbee: Error initializing RESET pin\n");
+            return -ENXIO;
+        }
+        gpio_set(reset_pin);
+    }
+    if (sleep_pin < GPIO_NUMOF) {
+        if (gpio_init_out(sleep_pin, GPIO_NOPULL) < 0) {
+            DEBUG("xbee: Error initializing SLEEP pin\n");
+            return -ENXIO;
+        }
+        gpio_clear(sleep_pin);
+    }
+    /* if reset pin is connected, do a hardware reset */
+    if (reset_pin < GPIO_NUMOF) {
+        gpio_clear(reset_pin);
+        hwtimer_wait(HWTIMER_TICKS(RESET_DELAY));
+        gpio_set(reset_pin);
+    }
+     /* put the XBee device into command mode */
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    _at_cmd(dev, "+++");
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    /* disable non IEEE802.15.4 extensions */
+    _at_cmd(dev, "ATMM2\r");
+    /* put XBee module in "API mode without escaped characters" */
+    _at_cmd(dev, "ATAP1\r");
+    /* apply AT commands */
+    _at_cmd(dev, "ATAC\r");
+    /* exit command mode */
+    _at_cmd(dev, "ATCN\r");
+
+    /* set default short address */
+    _set_addr(dev, dev->addr_short, 2);
+    /* load long address (we can not set it, as it is read only) */
+    _get_addr_long(dev, dev->addr_long, 8);
+    /* set default channel */
+    tmp[1] = 0;
+    tmp[0] = XBEE_DEFAULT_CHANNEL;
+    _set_channel(dev, tmp, 2);
+    /* set default PAN ID */
+    tmp[1] = (uint8_t)(XBEE_DEFAULT_PANID >> 8);
+    tmp[0] = (uint8_t)(XBEE_DEFAULT_PANID & 0xff);
+    _set_panid(dev, tmp, 2);
+
+    DEBUG("xbee: Initialization successful\n");
+    return 0;
+}
+
+int _send(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
+{
+    xbee_t *dev = (xbee_t *)netdev;
+    size_t size;
+    size_t pos;
+    ng_netif_hdr_t *hdr;
+    ng_pktsnip_t *payload;
+
+    /* check device descriptor and packet */
+    if (pkt == NULL) {
+        return -ENOMSG;
+    }
+    if (dev == NULL) {
+        ng_pktbuf_release(pkt);
+        return -ENODEV;
+    }
+
+    /* figure out the size of the payload to send */
+    size = ng_pkt_len(pkt->next);
+    if (size > XBEE_MAX_PAYLOAD_LENGTH) {
+        DEBUG("xbee: Error sending data, payload length exceeds limit\n");
+        ng_pktbuf_release(pkt);
+        return 0;
```

Comment:
would return `-EBADMSG` or `-ECOMM` or at least something negative (I think I used `-EBADMSG` in `ng_tapnet`).